zanata-war/src/main/java/org/zanata/service/impl/RequestServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+import org.zanata.model.HLocale;
+import org.zanata.model.LanguageRequest;
+import org.zanata.model.Request;
+import org.zanata.model.type.RequestState;
+import org.zanata.model.type.RequestType;
+import org.zanata.service.RequestService;
+import org.zanata.util.Event;
+
+import javax.persistence.EntityNotFoundException;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * @author Alex Eng <a href="aeng@redhat.com">aeng@redhat.com</a>
+ */
+@Name("requestServiceImpl")
+@Scope(ScopeType.STATELESS)
+public class RequestServiceImpl implements RequestService {
+
+    @In
+    private RequestDAO requestDAO;
+
+    @In
+    private LanguageRequestDAO languageRequestDAO;
+
+    @In("event")
+    private Event<RequestUpdatedEvent> requestUpdatedEvent;
+
+    @Override
+    public LanguageRequest createLanguageRequest(HAccount requester,
+            HLocale locale, boolean isRequestAsCoordinator,
+            boolean isRequestAsReviewer, boolean isRequestAsTranslator)
+            throws RequestExistException {
+        //search if there's any existing language request
+        //of the same requester, account and locale
+
+        LanguageRequest languageRequest =
+            languageRequestDAO.findRequesterPendingRequests(requester,
+                locale.getLocaleId());
+
+        if (languageRequest != null) {
+            throw new RequestExistException(
+                RequestType.LOCALE + ", " + requester.getUsername() + ", " +
+                    ", " + locale.getDisplayName());
+        }
+        String entityId = UUID.randomUUID().toString();
+        Request request = new Request(RequestType.LOCALE, requester, entityId);
+
+        languageRequest =
+            new LanguageRequest(request, locale, isRequestAsCoordinator,
+                isRequestAsReviewer, isRequestAsTranslator);
+        requestDAO.makePersistent(request);
+        languageRequestDAO.makePersistent(languageRequest);
+        return languageRequest;
+    }
+
+    @Override
+    public boolean isLanguageRequestExist(HAccount requester, HLocale locale) {
+        LanguageRequest languageRequest =
+                languageRequestDAO.findRequesterPendingRequests(requester,
+                    locale.getLocaleId());
+        return languageRequest != null;
+    }
+
+    @Override
+    public void updateLanguageRequest(Long requestId, HAccount actor,
+        RequestState state, String comment) throws EntityNotFoundException {
+        LanguageRequest languageRequest = languageRequestDAO.findById(requestId);
+        if (languageRequest != null) {
+            Request request = languageRequest.getRequest();
+
+            Request newRequest = request.update(actor, state, comment);
+            requestDAO.makePersistent(request);
+
+            //Need to add 1 second to request validTo date to avoid unique index constraint
+            newRequest.setValidTo(addOneSecond(request.getValidTo()));
```

Comment:
If we set newRequest.validFrom to oldRequest.validTo, and leave newRequest.validTo as null, we shouldn't need to do this.